### PR TITLE
~~STOP! AK TIEM!~~ Lights! Camera! AKtion!

### PR DESCRIPTION
**Pull request by alimpfard**
- ~~Depends on #1281 ~~
- ~~There are some changes in AK, depends on those too ([Serenity PR](https://github.com/SerenityOS/serenity/pull/16396))~~ :marge:'d
- Stage 2 success~
- ~~TODO: Clean this mess up lol~~

Unimplemented / boogs: NOTHING, CERTIFIED BOOG FREE :tm:
- ~~is~~
- ~~%=~~
- ~~ResultType in ErrorOr~~
- ~~make OptionalNone work with WeakPtr~~
- ~~lazy_emplace~~
- ~~whatever is happening to missing_return_type_arrow.jakt~~
- ~~whatever is going on in simple_match~~
- ~~WeakPtr::value()~~
- ~~type mismatches between AK/Jakt (String::split(), at least)~~ (cheated)

Stuff to do:
- [x] Figure out what to do with the bootstrap's copy of AK
- [x] Clean this up
- [x] windows
- [x] building with `jakt` itself (will unbork jakttest too)
- [X] Thank jt for the super ultra amazing new PR title.

### CODE REVIEW

File: runtime/CMakeLists.txt [hunk 1]
```diff
@@ -1,13 +1,73 @@
...
+    )
+endif()
+
+FetchContent_GetProperties(Serenity)
+if (NOT serenity_POPULATED)
+    FetchContent_Populate(Serenity)
+endif()
+
+file(REMOVE_RECURSE ${CMAKE_CURRENT_SOURCE_DIR}/AK)
+file(COPY ${serenity_SOURCE_DIR}/AK DESTINATION ${CMAKE_CURRENT_SOURCE_DIR})
```
[the remark this answers is not in the document]
Comment (ADKaster) [reply runtime/CMakeLists.txt:28]: Oh right. Wait why is build interface not public includes on 77? Bootstrap reasons?

File: runtime/IO/File.cpp [hunk 2]
```diff
@@ -37,9 +39,10 @@ File::~File()
     fclose(m_stdio_file);
 }
 
-ErrorOr<NonnullRefPtr<File>> File::open_for_reading(String path)
+ErrorOr<NonnullRefPtr<File>> File::open_for_reading(StringView path_)
 {
-    auto* stdio_file = fopen(path.c_string(), "rb");
+    auto path = DeprecatedString(path_);
+    auto* stdio_file = fopen(path.characters(), "rb");
```
Comment (ADKaster) [inline runtime/IO/File.cpp:45]: Pre existing but this leaks the stdio file on TRY error path

File: cmake/jakt-executable.cmake [hunk 3]
```diff
@@ -44,6 +47,8 @@ function(add_jakt_compiler_flags target)
     endif()
     set_property(TARGET "${target}" PROPERTY MSVC_RUNTIME_LIBRARY "MultiThreaded$<$<CONFIG:Debug>:Debug>")
   endif()
+
+  target_include_directories("${target}" BEFORE PRIVATE "${JAKT_AK_SOURCE_DIR}" "${JAKT_AK_BINARY_DIR}")
```
[the remark this answers is not in the document]
Comment (ADKaster) [reply cmake/jakt-executable.cmake:51]: Wait what? I must have missed it. Where are you installing AK headers +cpp files + AK/Debug.h into INSTALL_INCLUDEDIR/runtime/ or w/e? We can't be referencing the build directory from the install tree ever.

**Review by ADKaster — COMMENTED** [review]
Phone post review, pls ignore psedocode. 

I'm confused about install story, how we get all AK files in right place with CMake not referencing build paths

File: CMakeLists.txt [hunk 4]
```diff
@@ -49,7 +49,18 @@ include(cmake/jakt-executable.cmake)
 add_subdirectory(runtime)
 
 file(GLOB JAKT_STAGE0_SOURCES CONFIGURE_DEPENDS "bootstrap/stage0/*.cpp")
-add_executable(jakt_stage0 "${JAKT_STAGE0_SOURCES}")
+file(GLOB JAKT_STAGE0_RUNTIME_SOURCES CONFIGURE_DEPENDS "bootstrap/stage0/runtime/*.cpp")
+file(GLOB JAKT_STAGE0_RUNTIME_LIB_SOURCES CONFIGURE_DEPENDS "bootstrap/stage0/runtime/*/*.cpp")
```
[the remark this answers is not in the document]
Comment (alimpfard) [reply CMakeLists.txt:53]: Why? I see zero reason to list these manually, they don't change, so the "globbing makes builds break when new files are added" argument is void
I _can_ do it the annoying way, I'm just not sure why this is sad

File: cmake/jakt-executable.cmake (hunk 3 above)
[the remark this answers is not in the document]
Comment (alimpfard) [reply cmake/jakt-executable.cmake:51]: AK is copied into the _source dir_, which is later picked up by the recursive copy done by install()
We're not referencing the build dir here.

File: runtime/CMakeLists.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (alimpfard) [reply runtime/CMakeLists.txt:28]: Changing AK from serenity's checkout indeed does nothing, re the other rebuild, might be the effect of cache, but I didn't notice a full rebuild when rerunning ninja.

Hard link could work I guess, no idea how though

File: cmake/jakt-executable.cmake (hunk 3 above)
[the remark this answers is not in the document]
Comment (ADKaster) [reply cmake/jakt-executable.cmake:51]: What do you mean? You're adding the jakt_ak_binary_dir here. 

Is the file `build/serenity-build/AK/Debug.h` being installed through that recursive copy?

[the remark this answers is not in the document]
Comment (alimpfard) [reply cmake/jakt-executable.cmake:51]: That variable doesn't actually exist anymore ^^'
The debug.h file is being put into to source tree too

File: jakttest/fs.cpp [hunk 5]
```diff
@@ -3,9 +3,11 @@
...
 // SPDX-License-Identifier: BSD-2-Clause
 #include "fs.h"
-#include <Jakt/RefPtr.h>
-#include <Jakt/String.h>
-#include <Jakt/StringBuilder.h>
+#include <Jakt/DeprecatedString.h>
+
+#include <AK/RefPtr.h>
+#include <AK/DeprecatedString.h>
+#include <Jakt/DeprecatedStringBuilder.h>
```
[the remark this answers is not in the document]
Comment (alimpfard) [reply jakttest/fs.cpp:10]: :soon:

File: runtime/CMakeLists.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (alimpfard) [reply runtime/CMakeLists.txt:28]: We can't trivially `add_subdirectory(AK)`, that depends on the main serenity CMakeList to be included too.
I'll let you play with this yourself as discussed on Discord :^)

**Pull request merged**